util/pyclient/bft_client.py

```diff
@@ -241,3 +247,257 @@ def get_rsi_replies(self):
+        super().__init__(config, replicas)
+        self.sock = trio.socket.socket(trio.socket.AF_INET, trio.socket.SOCK_DGRAM)
+        self.port = bft_network.data_port_from_node_id(self.client_id)
+
+    async def _comm_prepare(self, read_only):
+        """ Bind the socket to 'localhost':port, where each port is a function of its client_id """
+        await self.sock.bind(('localhost', self.port))
+        self.comm_prepared = True
+
+    async def on_started_replicas(self, replica_ids):
+        """ UDP do nothing when replicas are started """
+        pass
+
+    async def on_stopped_replicas(self, replica_ids):
+        """ UDP do nothing when replicas are stopped """
+        pass
+
+    async def _send_data(self, data, replica):
+        """Send data. This operation always succeed after blocking, so True is always returned."""
+        await self.sock.sendto(data, (replica.ip, replica.port))
+        return True
+
+    async def _recv_data(self, required_replies, dest_replicas, cancel_scope):
+        """ Receive reply messages until a quorum is achieved or the enclosing cancel_scope times out. """
+        replicas_addr = [(r.ip, r.port) for r in dest_replicas]
+        while True:
+            data, sender = await self.sock.recvfrom(self.config.max_msg_size)
+            self._process_received_msg(data, sender, replicas_addr, required_replies, cancel_scope)
+
+    async def __aenter__(self):
+        """async context manager method for 'with' statements"""
+        return self
+
+    async def __aexit__(self, *args):
+        """async context manager method for 'with' statements"""
+        self.sock.close()
+
+    def __enter__(self):
+        """context manager method for 'with' statements"""
+        return self
+
+    def __exit__(self, *args):
+        """context manager method for 'with' statements"""
+        self.sock.close()
+        pass
+
+    def _process_received_msg_err(self, ex):
+        """In UDP, we do not except any errors while processing messages"""
+        # re-raise exception
+        raise ex
+
+class TcpTlsClient(BftClient):
+    """
+    Define a TCP/TLS client. This client communicates as a TLS over TCP client to all replicas.
+    It uses TCP to guarantee in-order data arrival and data reliability (using re-transmissions) and TLS to guarantee
+    authenticity and integrity. To enable a TLS-handshake, it uses self-signed SSL X.509 certificates.
+    """
+    # In create_tls_certs.sh - openssl command line utility uses CN(certificate name) in the subj field.
+    # This is the host name (domain name) to be verified.
+    CERT_DOMAIN_FORMAT="node%dser"
+    # Taken from TlsTCPCommunication.cpp (we prefer hard-code and not to parse the file)
+    MSG_HEADER_SIZE=4
+
+    def __init__(self, config, replicas, bft_network):
+        super().__init__(config, replicas)
+        self.ssl_streams = dict()
+        self.bft_network = bft_network
+
+    async def on_started_replicas(self, replica_ids):
```

Comment:
I'm not sure this callback really makes sense for a general client. It works when we only use the client for testing, but that's not the only use case. This client is a general client that should be usable with a production system. It's certainly not perfect, and the retry strategy can be improved, but I wouldn't want to limit it simply to being test only.

A better way to do this would be to do what we do in the C++ client and take a configuration, and continuously try to reconnect any broken streams in the background when detected. If a stream isn't currently connected, messages get dropped, but we are constantly trying to reconnect.

Reply:
@andrewjstone / @teoparvanov 
I think it related to your previous comment.
Trying to reconnect all the time slows down the test.
Just to understand the number of connections we have in a single test:
32 clients (30 regular and 2 reserved) with 7 replicas ~ 500 connections (SSL handshake) on a single machine every time a network is going up. I've tried to reduce the number of clients, but some tests failed (not adaptable to client number, making assumptions on the number of sends in a time interval, and expect moving from optimizing the fast path to slow path) so for now, returned to 30 clients.
Each replica that goes down triggers all client that is sending to get busy with reconnecting, and the send command would not end until a connection is failed (if the replica is still isolated/down).

THE MAJOR with your suggestion my opinion is that the current test framework implementation does not support this kind of implementation:
We are using blocking I/O on a single thread to deal with everything and avoid multi-threading.
To do what you suggest requires (at least) 1 thread per client that tries to reconnect the replica all the time in the background. for now, we can try to reconnect only during sending, and as I've said that causes many issues in test timeouts.